lib/trebuchet.rb

```diff
@@ -34,30 +32,57 @@ def set_backend(backend_type, *args)
     # this only works with additional args, e.g.: Trebuchet.backend = :memory
     alias_method :backend=, :set_backend
 
+    alias_method :threadsafe_state?, :threadsafe_state
+
     # Logging done at class level
     # TODO: split by user identifier so instance can return scoped to one user
     # (in case multiple users have user.trebuchet called)
     def initialize_logs
-      @logs = {}
+      state.logs = {}
     end
 
     def log(feature_name, result)
-      initialize_logs if @logs == nil
-      @logs[feature_name] = result
+      initialize_logs if state.logs.nil?
+      logs[feature_name] = result
+    end
+
+    def logs
+      state.logs
     end
 
-    attr_reader :logs
-    attr_accessor :current
+    def current=(other)
+      state.current = other
+    end
 
     def current
-      @current ||= @current_block.call if @current_block.respond_to?(:call)
-      @current || new(nil) # return an blank Trebuchet instance if @current is not set
+      state.current ||= current_block.call if current_block.respond_to?(:call)
+      state.current || new(nil) # return an blank Trebuchet instance if @current is not set
     end
 
     def reset_current!
-      @current = nil
+      self.current = nil
+    end
+
+    def thread_local_key
+      :trebuchet_state
     end
 
+    # state is a representation of the current context of Trebuchet such as current and
+    # current_proc, which are expected to be different between threads or fibers.
+    # exception_handler and backend are not included in this state object as they are
+    # not expected to change from fiber to fiber or request to request, therefore they
+    # must be thread/fibersafe on their own accord.
+    def state
```

Comment:
In https://github.com/socketry/async/pull/52 I implemented a way to copy state from one fiber to another, but this would require adding a `state=` setter to Trebuchet.

However, that PR hasn't been accepted yet so if you want, we could wait before adding a setter.

Reply:
Setter added!

Reply:
I had been exploring how to support the copying of state from one thread to another but was figuring that it would have to be done manually by the code that forks the thread setting the thread local itself. Hence the exposure of `thread_local_key`. I like this *much* better.